lib/Transforms/CombStrengthReduction.cpp

```diff
@@ -0,0 +1,252 @@
+//===- CombStrengthReduction.cpp - A Combinational Strength Reduction pass ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a simplified canonicalizer pass that doesn't do CFG
+// optimizations and other things that aren't helpful for many hardware IRs.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Transforms/Passes.h"
+
+#include "circt/Support/LLVM.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Rewrite/PatternApplicator.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace circt;
+using namespace comb;
+
+#define GEN_PASS_CLASSES
+#include "circt/Transforms/Passes.h.inc"
+
+static std::pair<size_t, size_t> computeCommonPrefixAndSuffix(const OperandRange &a, const OperandRange &b)
+{
+  size_t commonPrefixLength = 0;
+  size_t commonSuffixLength = 0;
+  size_t sizeA = a.size();
+  size_t sizeB = b.size();
+
+  for (; commonPrefixLength < std::min(sizeA, sizeB); commonPrefixLength++) {
+    if (a[commonPrefixLength] != b[commonPrefixLength]) {
+      break;
+    }
+  }
+
+  for (; commonSuffixLength < std::min(sizeA, sizeB) - commonPrefixLength; commonSuffixLength++) {
+    if (a[sizeA - commonSuffixLength - 1] != b[sizeB - commonSuffixLength - 1]) {
+      break;
+    }
+  }
+
+  return { commonPrefixLength, commonSuffixLength };
+}
+
+static bool isPredicateSigned (const ICmpPredicate & predicate)
+{
+  switch (predicate)  {
+    case ICmpPredicate::ult:
+    case ICmpPredicate::ugt:
+    case ICmpPredicate::ule:
+    case ICmpPredicate::uge:
+    case ICmpPredicate::ne:
+    case ICmpPredicate::eq:
+      return false;
+    case ICmpPredicate::slt:
+    case ICmpPredicate::sgt:
+    case ICmpPredicate::sle:
+    case ICmpPredicate::sge:
+      return true;
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+static mlir::Optional<Value> findFirstNonEmptyValue(const OperandRange &range)
+{
+  for (auto op : range) {
+    if (op.getType().getIntOrFloatBitWidth() > 0) {
+      return mlir::Optional<Value>(op);
+    }
+  }
+
+  return mlir::None;
```

Comment:
I don't think you need mlir:: qualifiers on None and Optional given you're including llvm.h.  Please trim them off

Reply:
done